src/ppm.rs

```diff
@@ -0,0 +1,499 @@
+    let mut msgs = Vec::with_capacity(num_shares);
+    msgs.push(Upload {
+        input_share: Share::Leader(leader_input_share),
+        proof_share: Share::Leader(leader_proof_share),
+        joint_rand_seed_hint: leader_joint_rand_seed_hint,
+        blind: leader_blind,
+    });
+
+    for helper in helper_shares.into_iter() {
+        msgs.push(Upload {
+            input_share: Share::Helper(helper.input_share, input_len),
+            proof_share: Share::Helper(helper.proof_share, proof_len),
+            joint_rand_seed_hint: helper.joint_rand_seed_hint,
+            blind: helper.blind,
+        });
+    }
+
+    Ok(msgs)
+}
+
+/// The message sent by an aggregator to every other aggregator. This is the final message of the
+/// input-validation protocol.
+#[derive(Clone, Debug)]
+pub struct Verifier<F: FieldElement> {
+    /// The aggregator's share of the verifier message.
+    pub verifier_share: Vec<F>,
+
+    /// The aggregator's share of the joint randomness, derived from its input share.
+    pub joint_rand_seed_share: Vec<u8>,
+}
+
+/// The initial state of an aggregate upon receiving the [`Upload`] message from the client.
+#[derive(Clone, Debug)]
+pub struct AggregatorState<F, V>
+where
+    F: FieldElement,
+    V: pcp::Value<F>,
+{
+    input_share: V,
+    verifier_share: Vec<F>,
+    joint_rand_seed_hint: Vec<u8>,
+}
+
+/// Run by each aggregator, this consumes the [`Upload`] message sent from the client and
+/// produces the [`Verifier`] message that will be broadcast to the other aggregators. It also
+/// returns the aggregator's state.
+///
+/// # Parameters
+///
+/// * `msg` is the message sent by the client. It contains, among other things, an input and proof
+/// share.
+/// * `param` is used to reconstruct the input share from the raw message.
+/// * `query_rand_seed` is used to derive the query randomness shared by all the aggregators.
+pub fn verify_start<F, V>(
+    msg: Upload<F>,
+    param: V::Param,
+    aggregator_id: u8,
+    query_rand_seed: &[u8],
+) -> Result<(AggregatorState<F, V>, Verifier<F>), PpmError>
+where
+    F: FieldElement,
+    V: pcp::Value<F>,
+{
+    assert_safe();
+
+    let input_share_data: Vec<F> = Vec::try_from(msg.input_share)?;
+    let input_share = V::try_from((param, &input_share_data))?;
+
+    let proof_share_data: Vec<F> = Vec::try_from(msg.proof_share)?;
+    let proof_share = pcp::Proof::from(proof_share_data);
+
+    // Compute the joint randomness.
+    let mut joint_rand_seed_share = vec![0; SEED_LENGTH];
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(&[aggregator_id]);
+    hasher.update(&msg.blind);
+    for x in input_share.as_slice() {
+        hasher.update(&(*x).into());
+    }
+    joint_rand_seed_share.copy_from_slice(hasher.finalize().as_bytes());
+
+    let mut joint_rand_seed = vec![0; SEED_LENGTH];
+    for j in 0..SEED_LENGTH {
+        joint_rand_seed[j] = msg.joint_rand_seed_hint[j] ^ joint_rand_seed_share[j];
+    }
+
```

Comment:
nit: use the `for (x, y) in joint_rand_seed.iter_mut().zip(joint_rand_seed_share.iter()) {}` form

Reply:
Doesn't quite work here because we have three variables instead of just two: joint_rand_seed, msg.joint_rand_seed_hint, and joint_rand_seed_share. I changed the iterator to `joint_rand_seed.iter_mut().enumerate()`, which is a bit more consistent.